lib/core/util.c

```diff
@@ -769,6 +769,117 @@ h2o_iovec_t h2o_build_destination(h2o_req_t *req, const char *prefix, size_t pre
     return h2o_concat_list(&req->pool, parts, num_parts);
 }
 
+#define SERVER_TIMING_DURATION_LONGEST_STR "dur=" H2O_INT32_LONGEST_STR ".000"
+
+size_t stringify_duration(char *buf, int64_t usec)
+{
+    int32_t msec = (int32_t)(usec / 1000);
+    usec -= ((int64_t)msec * 1000);
+    char *pos = buf;
+    pos += sprintf(pos, "dur=%" PRId32, msec);
+    if (usec != 0) {
+        *pos++ = '.';
+        int denom;
+        for (denom = 100; denom != 0; denom /= 10) {
+            int d = (int)usec / denom;
+            *pos++ = '0' + d;
+            usec -= d * denom;
+            if (usec == 0)
+                break;
+        }
+    }
+    return pos - buf;
+}
+
+#define DELIMITER ", "
+#define ELEMENT_LONGEST_STR(name) #name "; " SERVER_TIMING_DURATION_LONGEST_STR
+#define EMIT_ELEMENT(req, dst, name, max_len) emit_server_timing_element(req, dst, #name, h2o_time_compute_##name##_time, max_len)
+
+static void emit_server_timing_element(h2o_req_t *req, h2o_iovec_t *dst, const char *name,
+                                       int (*compute_func)(h2o_req_t *, int64_t *), size_t max_len)
+{
+    int64_t usec;
+    if (compute_func(req, &usec) == 0)
+        return;
+    if (dst->len == 0) {
+        if (max_len != SIZE_MAX)
+            dst->base = h2o_mem_alloc_pool(&req->pool, *dst->base, max_len);
+    } else {
+        dst->base[dst->len++] = ',';
+        dst->base[dst->len++] = ' ';
+    }
+    size_t name_len = strlen(name);
+    memcpy(dst->base + dst->len, name, name_len);
+    dst->len += name_len;
+    dst->base[dst->len++] = ';';
+    dst->base[dst->len++] = ' ';
+    dst->len += stringify_duration(dst->base + dst->len, usec);
+}
+
+void h2o_add_server_timing_header(h2o_req_t *req)
+{
+    /* caller needs to make sure that trailers can be used */
+    if (0x101 <= req->version && req->version < 0x200)
+        assert(req->content_length == SIZE_MAX);
+
+    /* add trailer header */
+    h2o_add_header_by_str(&req->pool, &req->res.headers, H2O_STRLIT("trailer"), 0, NULL, H2O_STRLIT("server-timing"));
+
+    /* emit timings */
+    h2o_iovec_t dst = {NULL};
+
+#define LONGEST_STR                                                                                                                \
+    ELEMENT_LONGEST_STR(connect)                                                                                                   \
+    DELIMITER ELEMENT_LONGEST_STR(header) DELIMITER ELEMENT_LONGEST_STR(body) DELIMITER ELEMENT_LONGEST_STR(request_total)         \
+        DELIMITER ELEMENT_LONGEST_STR(process) DELIMITER ELEMENT_LONGEST_STR(response)
+    size_t max_len = sizeof(LONGEST_STR);
+
+    EMIT_ELEMENT(req, &dst, connect, max_len);
+    EMIT_ELEMENT(req, &dst, header, max_len);
+    EMIT_ELEMENT(req, &dst, body, max_len);
+    EMIT_ELEMENT(req, &dst, request_total, max_len);
+    EMIT_ELEMENT(req, &dst, process, max_len);
```

Comment:
I believe that the names should match those defined for the Access Timings extensions defined for the access-log handler, with the `-time` suffix being omitted.

In other words, how about:
1. changing the names to be: connect, request-header, request-body, request-total, process, response
2. define `total-time` as a synonym of `duration` in the access-log handler

2 ensures us that we can continue to emit `server-timing: total; dur=xxx` as the trailer.

Reply:
Thank you for your comment. I addressed 1 at https://github.com/h2o/h2o/pull/1646/commits/51125dcc6cf206de18e3d5e24f1219f55fab71d0 and 2 at https://github.com/h2o/h2o/pull/1646/commits/3a8f262078203321163386685f68c72b47df4c77